### PR TITLE
Feature/types

### PR DESCRIPTION
**Pull request by RyanKirby89**
*Create Types for all Services*

Create definitions for each object that EPCC owns based on functions (Function.Products), need interfaces (product, etc)

Need:

definitions for every object that EPCC owns and is available in SDK
https://github.com/moltin/js-sdk/tree/master/src/endpoints

Create any new definitions for objects not defined in SDK (evaluate if needed first).
Compare endpoints in SDK and see if any are missing based on APIs on EPCC APIs. Then add new endpoints.

### CODE REVIEW

File: src/types/order.d.ts [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      created_at?: Date
+      updated_at?: Date
+    },
+    le?: {
+      with_tax?: number
+      without_tax?: number
+      created_at?: Date
+      updated_at?: Date
+    }
+
```
Comment (RyanKirby89) [inline src/types/order.d.ts:124]: remove n/l

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
```
Comment (RyanKirby89) [inline .idea/.gitignore:1]: remove .idea and add to git_ignore

File: src/types/cart.d.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+/**
+ * Carts
+ * Description: A Cart contains the product and custom cart items that a user may wish to purchase. Once a Cart is ready for Checkout, you can use the Checkout endpoint to convert the cart to an order.
```
Comment (RyanKirby89) [inline src/types/cart.d.ts:3]: needs n/l

File: src/types/collection.d.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        data: {
+          type: string
+          id: string
+        } []
+      }
+    }
+  }
+
+  /**
+   * DOCS:
```
Comment (RyanKirby89) [inline src/types/collection.d.ts:40]: docs?

File: src/types/currencies.d.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+   */
+  export interface CurrencyBase {
+    id?: string
+    type: string
+    code: string
+    exchange_rate: number // float
+    format: string
+    decimal_point: string
+    thousand_separator: string
+    decimal_places: number // int
```
Comment (RyanKirby89) [inline src/types/currencies.d.ts:23]: remove comments

File: src/types/customer.d.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * Core Product Base Interface
+   * For custom flows, extend this interface
+   * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/catalog/products/index.html
+   */
+  export interface CustomerBase {
+    id?:	string
+    type:	string
+    name:	string
```
Comment (RyanKirby89) [inline src/types/customer.d.ts:25]: spacing issues

File: src/types/customer.d.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/customers/index.html
+ */
+import { core } from './core'
+
+export as namespace customer
+
+export namespace customer {
+
+  /**
+   * Core Product Base Interface
```
Comment (RyanKirby89) [inline src/types/customer.d.ts:18]: Products?

File: src/types/customer.d.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+   * For custom flows, extend this interface
+   * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/catalog/products/index.html
```
Comment (RyanKirby89) [inline src/types/customer.d.ts:20]: docs to products?

File: src/types/field.d.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Fields
+ * Description: A Field represents a single Field of data (for example a Product Rating) to be applied to an entity. All Fields have a field_type (string, integer, boolean, date or relationship), a default value and an optional set of validation rules.
```
Comment (RyanKirby89) [inline src/types/field.d.ts:3]: n/l

File: src/types/field.d.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    name: string
+    slug: string
+    field_type: string
+    validation_rules: any[]
+    description: string
+    required: boolean
+    default: any
+    enabled: boolean
+    order: number
+    relationships: object
```
Comment (RyanKirby89) [inline src/types/field.d.ts:28]: Object? need to create proper interface for this

File: src/types/field.d.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  /**
+   * Field Endpoints
+   * DOCS:
+   * Get DOCS:
+   * Get All DOCS:
+   * Create DOCS:
+   * Update DOCS:
+   * Delete DOCS:
```
Comment (RyanKirby89) [inline src/types/field.d.ts:39]: Update docs

File: src/types/field.d.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+   */
+  export interface FieldsEndpoint extends core.CrudQueryableResource<FieldBase, {}, {}, {}> {
```
Comment (RyanKirby89) [inline src/types/field.d.ts:41]: no `{}`

File: src/types/file.d.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+ */
+import { core } from './core'
+
+export as namespace file
+
+export namespace file {
+  /**
+   * Core File Base Interface
+   * For custom flows, extend this interface
+   * DOCS:
```
Comment (RyanKirby89) [inline src/types/file.d.ts:14]: docs needed

File: src/types/file.d.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        height: number
+      }
+    }
+    links: {
+      self: string
+    }
+  }
+
+  /**
+   * DOCS:
```
Comment (RyanKirby89) [inline src/types/file.d.ts:41]: docs

File: src/types/file.d.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  /**
+   * File Endpoints
+   * DOCS:
+   * Get DOCS:
+   * Get All DOCS:
+   * Create DOCS:
+   * Update DOCS:
+   * Delete DOCS:
```
Comment (RyanKirby89) [inline src/types/file.d.ts:81]: docs

File: src/types/file.d.ts [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  export interface FileEndpoint extends core.CrudQueryableResource<FileBase, FileFilter, {}, {}> {
```
Comment (RyanKirby89) [inline src/types/file.d.ts:83]: no `{}`

File: src/types/flow.d.ts [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+ * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/advanced/custom-data/flows.html
+ */
+import { core } from './core'
+
+export as namespace flow
+
+export namespace flow {
+  /**
+   * Core Flow Base Interface
+   * DOCS:
```
Comment (RyanKirby89) [inline src/types/flow.d.ts:14]: docs

File: src/types/flow.d.ts [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    type: string
+    name: string
+    description: string
+    enabled: boolean
+  }
+
+  /**
+   * DOCS:
+   */
+  export interface FlowFilter {
```
Comment (RyanKirby89) [inline src/types/flow.d.ts:28]: can you filter flow fields?

File: src/types/intergration.d.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+   * Get DOCS:
+   * Get All DOCS:
+   * Create DOCS:
+   * Update DOCS:
+   * Delete DOCS:
+   */
+export interface IntegrationEndpoint extends core.CrudQueryableResource<IntegrationBase, IntegrationFilter, IntegrationSort, IntegrationInclude> {
+    endpoint: 'integration'
+  }
+}
```
Comment (RyanKirby89) [inline src/types/intergration.d.ts:42]: make plan for this

File: src/types/inventory.d.ts [hunk 20]
```diff
@@ -0,0 +1,36 @@
+/**
+ * Inventory
+ * Description: The Inventory API allows you to manage stock for products in your project catalog. Each product keeps a history of inventory transactions, enabling easier stock auditing.
```
Comment (RyanKirby89) [inline src/types/inventory.d.ts:3]: n\l

File: src/types/inventory.d.ts [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    total: number
+    available: number
+    allocated: number
+  }
+
+  /**
+   * Inventory Endpoints
+   * TODO Need to check on actually functionality
+   */
+  export interface InventoryEndpoint extends core.QueryableResource<InventoryBase, {}, {}, {}> {
```
Comment (RyanKirby89) [inline src/types/inventory.d.ts:27]: no `{}`

File: src/types/job.d.ts [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+     * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/advanced/jobs/get-all-jobs.html
+     */
+    All(): Promise<ResourcePage<JobBase>>
+
+    /**
+     * Title: Get
+     * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/advanced/jobs/get-a-job.html
+     */
+    Get(id: string): Promise<JobBase>
+
```
Comment (RyanKirby89) [inline src/types/job.d.ts:60]: remove n/l

File: src/types/order.d.ts [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+        }
+        timestamps: {
+          created_at: string
+          updated_at: string
+        }
+      }
+    }
+    relationships?: {
+      main_image?: Relationship<'product'> []
+      categories?: Relationship<'customer'>
```
Comment (RyanKirby89) [inline src/types/order.d.ts:70]: Can you do this for the other relationships...

File: src/types/order.d.ts [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+      amount: number
+      currency: string
+      includes_tax: boolean
+    }
+    value: {
+      amount: number
+      currency: string
+      includes_tax: boolean
+    }
+    links: {} // TODO - docs need updating
```
Comment (RyanKirby89) [inline src/types/order.d.ts:147]: dont forget this

File: src/types/order.d.ts [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+          type: string
+        }[]
+      }
+      taxes: {
+        data: {
+          id: string
+          type: string
+        }[]
+      }
+    }
```
Comment (RyanKirby89) [inline src/types/order.d.ts:205]: relationships - see products

File: src/types/product.d.ts [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+   * Product Endpoints
+   * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/products/index.html
+   * Get DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/catalog/products/get-a-product.html
+   * Get All DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/catalog/products/get-all-products.html
+   * Update DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/catalog/products/update-a-product.html
+   */
+  export interface ProductsEndpoint<T extends ProductBase> extends core.CrudQueryableResource<ProductBase | T, ProductFilter, ProductSort, ProductInclude> {
+    endpoint: 'products'
+
+    // TODO CREATE - UPDATE - DELETE
```
Comment (RyanKirby89) [inline src/types/product.d.ts:123]: remove this - this is done

File: .gitignore [hunk 27]
```diff
@@ -2,5 +2,6 @@ node_modules
 localStorage/
 npm-debug.log
 .npmrc
+.idea/
```
Comment (samblacklock) [inline .gitignore:5]: mebs this could go in global .gitignore instead of the repos? what do you think?

File: src/types/address.d.ts [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+ * A Customer may have multiple addresses, such as home, work and neighbor. You are able to use a customer with an
+ * implicitBearer token. This is recommend for client-side interactions. Or you can use a client_credentials Bearer
+ * token. This is recommended for back end interactions.
+ * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/index.html
+ */
+import { core, Relationship } from './core'
+
+export as namespace address
+
+export namespace address {
```
Comment (samblacklock) [inline src/types/address.d.ts:13]: whats the difference between `export as namespace` and `export namespace`?

File: src/types/cart.d.ts [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+    RemoveItem(itemId: string): Promise<ResourcePage<CartItem>>
+
+    UpdateItem(
+      itemId: string,
+      quantity: number,
+      data?: any
+    ): Promise<ResourcePage<CartItem>>
+
+    /**
+     * @deprecated Use UpdateItem method
```
Comment (samblacklock) [inline src/types/cart.d.ts:114]: are we just saying `UpdateItem` is a nicer name for this method? afaik the only update you can do on a product is quantity

File: src/types/cart.d.ts [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+
+
+    AddPromotion(code: string): Promise<ResourcePage<CartItem>>
+
+    RemoveItem(itemId: string): Promise<ResourcePage<CartItem>>
+
+    UpdateItem(
+      itemId: string,
+      quantity: number,
+      data?: any
```
Comment (samblacklock) [inline src/types/cart.d.ts:110]: looks like update takes `quantity` , `id`, and `type` - unless I am misunderstanding this 😆

File: src/types/collection.d.ts [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+ */
+import { core } from './core'
+
+export as namespace collection
+
+export namespace collection {
+  /**
+   * Core Collection Base Interface
+   * For custom flows, extend this interface
+   * DOCS:
```
Comment (samblacklock) [inline src/types/collection.d.ts:14]: needs a link 👀

File: src/types/customer.d.ts [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/customers/filtering.html
+   */
+  export interface CustomerFilter {
+    eq?: {
+      email?: string
+    }
+  }
+
+  type CustomerSort = '' // todo... need to check
```
Comment (samblacklock) [inline src/types/customer.d.ts:47]: still todo?

File: src/types/field.d.ts [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+    name: string
+    slug: string
+    field_type: string
+    validation_rules: any[]
+    description: string
+    required: boolean
+    default: any
+    enabled: boolean
+    order: number
+    relationships: Relationship<'addresses' | 'products' | 'brands' | 'collections' | 'categories' | 'customers' |
```
Comment (samblacklock) [inline src/types/field.d.ts:30]: I think this can also be a relationship to one other flow or many other flows... 🤯

File: src/types/file.d.ts [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    file_name: string
+    mime_type: string
+    file_size: number
+    public: boolean
+    meta?: {
+      timestamps: {
+        created_at: string
+      }
+      dimensions: {
```
Comment (samblacklock) [inline src/types/file.d.ts:30]: didn't know this was a thing 😄

File: src/types/flow.d.ts (hunk 18 above)
Comment (samblacklock) [reply src/types/flow.d.ts:24]: No 😢

File: .gitignore (hunk 27 above)
Comment (RyanKirby89) [reply .gitignore:5]: yeah, good shout

File: src/types/cart.d.ts (hunk 29 above)
Comment (RyanKirby89) [reply src/types/cart.d.ts:114]: this was actually in there already, I didnt want to change it haha

File: src/types/address.d.ts (hunk 28 above)
Comment (RyanKirby89) [reply src/types/address.d.ts:13]: a .d.ts files requires a export as namespace - not 100% sure why though. haha

File: src/types/cart.d.ts (hunk 30 above)
Comment (RyanKirby89) [reply src/types/cart.d.ts:110]: also, a dirty fat any stuck in there as well. Carts was pretty big, so I was thinking of trying to break this into a separate ticket to ensure its done properly, but will update this interface to ensure its correct. thanks.

File: .gitignore (hunk 27 above)
Comment (RyanKirby89) [reply .gitignore:5]: @samblacklock how do I actually include this in the global ignore?

File: src/types/cart.d.ts (hunk 30 above)
Comment (RyanKirby89) [reply src/types/cart.d.ts:110]: I actually think I just copied and pasted this function as well. Good spot, will update

File: src/types/collection.d.ts (hunk 31 above)
Comment (RyanKirby89) [reply src/types/collection.d.ts:14]: done

File: .gitignore (hunk 27 above)
Comment (samblacklock) [reply .gitignore:5]: @RyanKirby89 found some info here 👌 https://blog.martinhujer.cz/dont-put-idea-vscode-directories-to-projects-gitignore/

File: src/types/address.d.ts [hunk 35]
```diff
@@ -0,0 +1,48 @@
...
+   * Get DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/get-an-address.html
+   * Get All DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/get-all-addresses.html
+   * Create DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/create-an-address.html
+   * Update DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/update-an-address.html
+   * Delete DOCS: https://documentation.elasticpath.com/commerce-cloud/docs/api/orders-and-customers/addresses/delete-an-address.html
+   */
+  export interface AddressesEndpoint extends core.CrudQueryableResource<AddressBase, null, null, null> {
+    endpoint: 'addresses'
+  }
+}
```
Comment (shaunmaharaj) [inline src/types/address.d.ts:48]: Do you see any build errors for missing trailing whitespace in your IDE? Normally this should get flagged through a linter

File: src/types/address.d.ts [hunk 36]
```diff
@@ -0,0 +1,48 @@
...
+    line_1: string
+    line_2: string
+    city: string
+    county: string
+    postcode: string
+    country: string
+  }
+
+  /**
+   * Address Endpoints
```
Comment (shaunmaharaj) [inline src/types/address.d.ts:37]: I love the structure of these docs! nice!!

File: src/types/core.d.ts [hunk 37]
```diff
@@ -0,0 +1,136 @@
...
+    application?: string
+    client_id: string
+    client_secret?: string
+    host: string
+    protocol: 'https'
+    version: 'v2'
+    currency?: string
+    custom_fetch?: Function
+    auth: {
+      expires: 3600
```
Comment (shaunmaharaj) [inline src/types/core.d.ts:95]: We set the expiry for this ourselves? Is 1 hour sufficient for our token expiration?

File: src/types/address.d.ts (hunk 35 above)
Comment (RyanKirby89) [reply src/types/address.d.ts:48]: mm, my linter seemed to pass?

File: src/types/core.d.ts (hunk 37 above)
Comment (RyanKirby89) [reply src/types/core.d.ts:95]: @shaunmaharaj I'm not sure to be honest. This was just existing code I copied from moltin.d.ts - maybe worth another ticket to explore properly?

File: src/types/file.d.ts (hunk 34 above)
Comment (shaunmaharaj) [reply src/types/file.d.ts:30]: This is very cool, I didn't know this was possible:
https://documentation.elasticpath.com/commerce-cloud/docs/api/advanced/files/index.html

File: src/types/core.d.ts (hunk 37 above)
Comment (RyanKirby89) [reply src/types/core.d.ts:95]: I have actually noticed that it is a hardcoded figure in the js itself - in config.js

Comment (samblacklock) [reply src/types/core.d.ts:95]: @shaunmaharaj pretty sure its 1hr in the backend too

File: src/types/address.d.ts (hunk 35 above)
Comment (samblacklock) [reply src/types/address.d.ts:48]: @shaunmaharaj @RyanKirby89 i _think_  the prettier config should fix that if you have prettier in your editor, but not sure

File: src/types/core.d.ts (hunk 37 above)
Comment (shaunmaharaj) [reply src/types/core.d.ts:95]: Okay cool, we can keep it to stay consistent for now and revisit if we have to later

**Pull request merged**